File: BinaryTreeInorderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
 //Morris Traversal!! very tricky
class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> result;
        if(root==0) return result;
        TreeNode* p;
        TreeNode* pre;
    	for(p=root;p!=0;) {
    		if(p->left==0) { result.push_back(p->val); p=p->right; continue; }
    
    		for(pre=p->left;pre->right!=0 && pre->right!=p;pre=pre->right) { }
    
    		if(pre->right==0)
    			{ pre->right=p; p=p->left; continue; }
    		else
    			{ pre->right=0; result.push_back(p->val); p=p->right; continue; }
        }
    }
};

File: srcr2/BinaryTreeInorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 // Morris Traversal!!
public class Solution {
    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        TreeNode p = root;
        TreeNode pre = null;
        while (p != null){
            if (p.left == null) {
                result.add(p.val);
                p = p.right;
            }
            else {
                pre = getPreNode(p);
                if (pre.right == null){
                    pre.right = p;
                    p = p.left;
                } else {
                    result.add(p.val);
                    pre.right = null;
                    p = p.right;
                }
            }
        }
        return result;
    }
    
    TreeNode getPreNode(TreeNode e){
        TreeNode pre = e.left;
        while (pre.right != null && pre.right != e) pre = pre.right;
        return pre;
    }
}
